include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file defines the type for model predictive Model Predictive
+/// Control (MPC) on linear system dynamics by converting the MPC optimization
+/// problem to a quadratic cost problem. The new problem is optimized using the
+/// Operator Splitting Quadratic Program (OSQP).
+///
+/// \brief MPC Controller
+//===----------------------------------------------------------------------===//
+
+#ifndef LDSCTRLEST_LDS_MPCCTRL_H
+#define LDSCTRLEST_LDS_MPCCTRL_H
+
+// namespace
+#include "lds.h"
+// system type
+#include "lds_sys.h"
+
+// osqp
+#include "osqp.h"
+#include "osqp_api_constants.h"
+
+namespace lds {
+
+template <typename System>
+class MpcController {
+  static_assert(std::is_base_of<lds::System, System>::value,
+                "System must be derived from lds::System");
+
+ public:
+  /**
+   * @brief     Constructs a new MpcController.
+   */
+  MpcController() = default;
+
+  /**
+   * @brief     Constructs a new MpcController.
+   *
+   * @param     sys   The system being controlled
+   * @param     u_lb  The lower bound of the control input
+   * @param     u_ub  The upper bound of the control input
+   *
+   * @tparam    System  The system type
+   */
+  MpcController(const System& sys, Vector u_lb, Vector u_ub);
+
+  /**
+   * @brief     Constructs a new MpcController.
+   *
+   * @param     sys   The system being controlled
+   * @param     u_lb  The lower bound of the control input
+   * @param     u_ub  The upper bound of the control input
+   *
+   * @tparam    System  The system type
+   */
+  MpcController(System&& sys, Vector u_lb, Vector u_ub);
+
+  /**
+   *
+   */
+  Vector Control(data_t t_sim, const Vector& x0, const Vector& u0,
+                 const Matrix& xr);
+
+  // getters
+
+  // setters
+  void set_control(Matrix Q, Matrix R, Matrix S, size_t N, size_t M) {
+    Q_ = Q;
+    // R_ = R; // Isn't used
+    S_ = S;
+    N_ = N;
+    M_ = M;
+
+    // Set up P matrix
+    Matrix Px   = arma::kron(Matrix(N_, N_, arma::fill::eye), Q_);
+    Matrix Pu1  = arma::kron(Matrix(M_, M_, arma::fill::eye), 2 * S_ + R);
+    Matrix Pu2  = arma::kron(Matrix(eye_offset(M)) + Matrix(eye_offset(M, 1)), -S_);
+    Matrix Pu3  = block_diag(Matrix((M_ - 1) * m_, (M_ - 1) * m_, arma::fill::zeros), -S_);
+    Matrix Pu = Pu1 + Pu2 + Pu3;
+    P_ = arma::trimatu(2 * block_diag(Px, Pu)); // Taking only the upper triangular part
```

Comment:
Are these sparse, like in the Python lqmpc? That P matrix gets pretty big, right?

Reply:
They aren't sparse right now, but I think I should be able to make them sparse. Speaking of sparse matrices, is it okay if I define a type for sparse matrices in `lds.h` like the `Matrix` and `Vector` types?

Reply:
Yep, I think it would be fine to define a sparse matrix type in lds.h